src/web-extensions.c

```diff
@@ -0,0 +1,794 @@
+ * License along with this library. If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include "config.h"
+
+#include <stdint.h>
+#include <sys/types.h>
+#include <sys/wait.h>
+#include <glib/gi18n.h>
+#include <gio/gunixfdlist.h>
+#include <json-glib/json-glib.h>
+
+#include "session.h"
+#include "web-extensions.h"
+#include "request.h"
+#include "permissions.h"
+#include "xdp-dbus.h"
+#include "xdp-impl-dbus.h"
+#include "xdp-utils.h"
+
+#define PERMISSION_TABLE "webextensions"
+
+typedef struct _WebExtensions WebExtensions;
+typedef struct _WebExtensionsClass WebExtensionsClass;
+
+struct _WebExtensions
+{
+  XdpDbusWebExtensionsSkeleton parent_instance;
+};
+
+struct _WebExtensionsClass
+{
+  XdpDbusWebExtensionsSkeletonClass parent_class;
+};
+
+static XdpDbusImplAccess *access_impl;
+static WebExtensions *web_extensions;
+
+GType web_extensions_get_type (void);
+static void web_extensions_iface_init (XdpDbusWebExtensionsIface *iface);
+
+G_DEFINE_TYPE_WITH_CODE (WebExtensions, web_extensions, XDP_DBUS_TYPE_WEB_EXTENSIONS_SKELETON,
+                         G_IMPLEMENT_INTERFACE (XDP_DBUS_TYPE_WEB_EXTENSIONS,
+                                                web_extensions_iface_init));
+
+typedef enum _WebExtensionsSessionMode
+{
+  WEB_EXTENSIONS_SESSION_MODE_CHROMIUM,
+  WEB_EXTENSIONS_SESSION_MODE_MOZILLA,
+} WebExtensionsSessionMode;
+
+typedef enum _WebExtensionsSessionState
+{
+  WEB_EXTENSIONS_SESSION_STATE_INIT,
+  WEB_EXTENSIONS_SESSION_STATE_STARTING,
+  WEB_EXTENSIONS_SESSION_STATE_STARTED,
+  WEB_EXTENSIONS_SESSION_STATE_CLOSED,
+} WebExtensionsSessionState;
+
+typedef struct _WebExtensionsSession
+{
+  Session parent;
+
+  WebExtensionsSessionMode mode;
+  WebExtensionsSessionState state;
+
+  GPid child_pid;
+  guint child_watch_id;
+
+  int standard_input;
+  int standard_output;
+  int standard_error;
+} WebExtensionsSession;
+
+typedef struct _WebExtensionsSessionClass
+{
+  SessionClass parent_class;
+} WebExtensionsSessionClass;
+
+GType web_extensions_session_get_type (void);
+
+G_DEFINE_TYPE (WebExtensionsSession, web_extensions_session, session_get_type ());
+
+static void
+web_extensions_session_init (WebExtensionsSession *session)
+{
+  session->child_pid = -1;
+  session->child_watch_id = 0;
+
+  session->standard_input = -1;
+  session->standard_output = -1;
+  session->standard_error = -1;
+}
+
+static void
+web_extensions_session_close (Session *session)
+{
+  WebExtensionsSession *web_extensions_session = (WebExtensionsSession *)session;
+
+  if (web_extensions_session->state == WEB_EXTENSIONS_SESSION_STATE_CLOSED) return;
+
+  web_extensions_session->state = WEB_EXTENSIONS_SESSION_STATE_CLOSED;
+  if (web_extensions_session->child_watch_id != 0)
+    {
+      g_source_remove (web_extensions_session->child_watch_id);
+      web_extensions_session->child_watch_id = 0;
+    }
+
+  if (web_extensions_session->child_pid > 0)
+    {
+      kill (web_extensions_session->child_pid, SIGKILL);
```

Comment:
@bandali0 What was described above is not implemented in the Firefox-side patch. To see the relevant code, use Ctrl+F "closeSession" at https://phabricator.services.mozilla.com/D140803. Are you going to implement it?

On the portal patch here, please add a comment that explains that the responsibility of graceful killing the application has been delegated to Firefox, and that the SIGKILL here is a last shot attempt to clean up if needed.

P.S. This is the third time I'm posting the comment, because for some reason it ended up in the wrong place when I posted before.

Reply:
I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1903309 as a follow-up on the Firefox side.

If we want graceful termination, then we'd need a new message in the protocol (to send SIGTERM), don't we?